### ISSUE TITLE
Keep errors while submiting to the server

### ISSUE THREAD

**Issue by andresgutgon** [IdoPesok/zsa#121]
# What?
Hi, I'm using `useServerAction` to save some data into my server. It's working fine but I see a glitch while `isPending`

The errors in the server are reset and the UI jumbs.

This is my form:
``` javascript
  const { data, executeFormAction, isPending, error, ...rest } = useServerAction(
    onboarding,
    {
      initialData: { name, username },
    },
  )
  const errorData = error?.data
  const fieldErrors = error?.fieldErrors
  return (
      <Input
        label='Your username'
        type='text'
        name='username'
        errors={fieldErrors?.username}
        defaultValue={data?.username ?? errorData?.username}
        onChange={(e) => setUserName(e.target.value)}
        placeholder='Pick a username'
        description={<UsernameDescription username={currentUsername} />}
      />
  )
```

This is my action:
``` javascript
export const onboarding = authProcedure
  .createServerAction()
  .input(input, { type: 'formData' })
  .output(input)
  .experimental_shapeError(({ err, typedData }) => {
    if (err instanceof ZSAError && err.code !== 'INPUT_PARSE_ERROR') {
      return {
        code: err.code,
        message: err.message,
      }
    }

    return {
      data: typedData?.inputRaw,
      fieldErrors: typedData?.inputParseErrors?.formattedErrors,
    }
  })
  .handler(async ({ input, ctx: { user } }) => {
    await validateUniqueUsername({ data: input, user })

    const result = await finishOnboarding({ user, data: input })
    const value = result.unwrap()
    await updateSession({
      user: { name: value.name, username: value.username },
    })

    redirect('/')
  })
```

This is my UI. Check the disappearing red errors when submitting

https://github.com/IdoPesok/zsa/assets/49499/b17d4941-0091-4758-a184-23fd6c254177

**Comment by andresgutgon** [comment]
Am I using it ok? Is there a way of avoiding this glitch?

**Comment by andresgutgon** [comment]
Also, to keep previous data, I had to use `shapeError` and get `typedData?.inputRaw`. How hard would always be to return data already present in the form? 

I say this because otherwise, all inputs are empty when one of them have a validation error no?

**Comment by IdoPesok** [comment]
Hi, thanks for bringing this up!

I understand the issue. Going to compare with other libraries and report back shortly with next steps.

In the meantime -- if you want to try -- you should be able to get this persistence with [useActionState](https://zsa.vercel.app/docs/forms#useactionstate). This also supports progressive enhancement.

> always be to return data already present in the form?

Do you mean to return `inputRaw` in a default TZSAError without needing to do shape error?

**Comment by andresgutgon** [comment]
> Do you mean to return inputRaw in a default TZSAError without needing to do shape error?

Yes!

**Comment by andresgutgon** [comment]
Using `state` gives me this TS error

I'm in latest `zsa`. In fact I'm debugging it locally with `npm link`
<img width="1245" alt="image" src="https://github.com/IdoPesok/zsa/assets/49499/62ce1c95-b10a-4adc-bc31-995d03b93ce8">

**Comment by IdoPesok** [comment]
What version of React is this?

**Comment by IdoPesok** [comment]
Also if you are using npm link, do you want to give the input schema functions a shot? Going to merge soon would be awesome if you can test that branch in your project and give feedback.

**Comment by andresgutgon** [comment]
> What version of React is this?

I'm living on the edge xD

``` json
{
    "next": "14.3.0-canary.59",
    "next-auth": "5.0.0-beta.19",
    "react": "19.0.0-beta-26f2496093-20240514",
    "react-dom": "19.0.0-beta-26f2496093-20240514"
}
```

**Comment by andresgutgon** [comment]
> Also if you are using npm link, do you want to give the input schema functions a shot?

I can try that. But would be ideal to use `useServerAction`

And out of the box get:

1. `data` persisted between attempts
2. Do not erase errors until the response come from the server (action)

But I think that would be ideal without needing to customize errors or other parts of the lib. It would make a straightforward understand of the model

**Comment by IdoPesok** [comment]
Ah yes, I meant outside of this issue and related to #120 so you don't need to do validate unique username in the handler.

**Comment by andresgutgon** [comment]
Ah!! yes yes I didn't understand you. Let me check that

**Comment by andresgutgon** [comment]
Uhm, I'm not sure I'm using in the right way. I'm doing this
<img width="624" alt="image" src="https://github.com/IdoPesok/zsa/assets/49499/981a53d2-fc53-4ce6-aad7-bf5126037b76">

But debugging I see this
<img width="731" alt="image" src="https://github.com/IdoPesok/zsa/assets/49499/e6571404-ed0b-42c1-b004-801b9daf2fa1">


Looks I'm not returning a `ZodType` I'm returning a zodobject
<img width="696" alt="image" src="https://github.com/IdoPesok/zsa/assets/49499/61328b97-949e-4b24-8366-85a8f81d5b99">

Is this wrong?

**Comment by andresgutgon** [comment]
If I force that my inputSchema is the one used
<img width="893" alt="image" src="https://github.com/IdoPesok/zsa/assets/49499/bfbdd783-1ead-43b0-8418-7906f8955a8a">


Then in the frontend it comes without `formattedErrors` it comes like a zod error
<img width="549" alt="image" src="https://github.com/IdoPesok/zsa/assets/49499/55f3bec1-17c5-4b12-a8b1-7708f664d510">

**Comment by andresgutgon** [comment]
Ok looking more into the code this makes the error appear on the form fields
<img width="1009" alt="image" src="https://github.com/IdoPesok/zsa/assets/49499/9e31a34a-0904-4631-9134-84f6df6c8aa6">


So if I understand correctly errors to be present not as a ZodError but as formatted errors needs to be processed by the final user (me)?

**Comment by IdoPesok** [comment]
Hi, the formatted errors come from [Zod formatted errors](https://zod.dev/ERROR_HANDLING?id=formatting-errors). Same thing with field errors, and form errors. These are standard Zod error schemas that people use. We can be more clear in our docs to explain this.

**Comment by andresgutgon** [comment]
I see, but my question is. For having formatted errors do I need to use `shapeError`. I'll try with stable branch but i think it's working without that method

**Comment by andresgutgon** [comment]
Doesn't matter for now anyway. 

My feedback with the PR for input function is this

https://github.com/IdoPesok/zsa/issues/121#issuecomment-2170083331

Basically I'm not sure what I have to pass in the `input`
I'm passing a `z.object` but is failing
```javascript
input(async ({ ctx }) => z.object(...), { type: 'formData'})
```
<img width="605" alt="image" src="https://github.com/IdoPesok/zsa/assets/49499/82bb2f61-47e3-4204-979c-35fc0e0c46dd">

**Comment by andresgutgon** [comment]
Yep, running tests in that PR it also fails
<img width="744" alt="image" src="https://github.com/IdoPesok/zsa/assets/49499/d266c1fa-9fd0-4877-bfad-48edb11cd163">

**Comment by andresgutgon** [comment]
Test are passing in your PR but I saw you left a `.only` here https://github.com/IdoPesok/zsa/commit/0a085123f15af9aa358ceabc80133909eb2054e9#diff-0003458ed2aa52f7ec50b708d43d47e5f345bd88e8ecf56495befcc271207f1eR909

**Comment by andresgutgon** [comment]
Sorry for all the spam 🙏 I'm trying to understand. 

Doing these 2 changes it works. No need for custom `shapeError`

Looks like something related with `instanceof` but no idea why these 2 return false

<img width="877" alt="image" src="https://github.com/IdoPesok/zsa/assets/49499/7345c496-9f51-4087-930f-416cb41b7d48">

**Comment by andresgutgon** [comment]
Maybe is related https://github.com/colinhacks/zod/issues/2241

**Comment by andresgutgon** [comment]
ALL_MY_FAULT 😂 

Oh man, NodeJS is hard. 
What happened is your monorepo use `NPM` and I was linking from my monorepo with `pnpm link ~/code/opensource/zsa/packages/zsa` and it was taking Zod from pnpm and so the instance was different when checking `instanceof` 

What I did instead is 

```json
{
  "zsa": "file:../../../../opensource/zsa/packages/zsa",
  "zsa-react": "file:../../../../opensource/zsa/packages/zsa-react",
}
```

Fuck, this was a good lesson. 

Your PR works great `input(function)` do what it promise 🎉

<img width="518" alt="image" src="https://github.com/IdoPesok/zsa/assets/49499/5541ec34-79de-4ab4-90f2-57373c535376">


I'll move now to investigate the problems in this PR

OFF-TOPIC:
What's the use case for `retry`. Reading the code it makes everything more complicated. I'm courios to know what made you implement retries in this package. I can't imagine any good cases

**Comment by andresgutgon** [comment]
> Polling and caching

Yep, these are frontend concerns and because zsa already integrates with react-query there's no need to bring that complexity to this code.

> Okay, I am going to move this conversation to https://github.com/IdoPesok/zsa/pull/124 as the code is there.

Nice, happy to change anything
